zebra-network/src/address_book.rs

```diff
@@ -214,129 +243,197 @@ impl AddressBook {
+    }
+
+    /// Return an iterator over peers that have recently sent us messages,
+    /// in arbitrary order.
+    fn recently_live_peers(&'_ self) -> impl Iterator<Item = MetaAddr> + '_ {
+        let _guard = self.span.enter();
 
-        // Skip live peers, and peers pending a reconnect attempt, then sort using BTreeSet
-        self.by_addr
-            .values()
-            .filter(move |peer| !self.maybe_connected_addr(&peer.addr))
-            .collect::<BTreeSet<_>>()
-            .into_iter()
+        self.peers_unordered()
+            .filter(move |peer| self.recently_live_addr(peer.addr))
             .cloned()
     }
 
-    /// Return an iterator over all the peers in `state`, in arbitrary order.
-    pub fn state_peers(&'_ self, state: PeerAddrState) -> impl Iterator<Item = MetaAddr> + '_ {
+    /// Return an iterator over peers that have recently failed,
+    /// in arbitrary order.
+    fn recently_failed_peers(&'_ self) -> impl Iterator<Item = MetaAddr> + '_ {
         let _guard = self.span.enter();
 
-        self.by_addr
-            .values()
-            .filter(move |peer| peer.last_connection_state == state)
+        self.peers_unordered()
+            .filter(move |peer| self.recently_failed_addr(peer.addr))
             .cloned()
     }
 
-    /// Return an iterator over peers that might be connected, in arbitrary
-    /// order.
-    pub fn maybe_connected_peers(&'_ self) -> impl Iterator<Item = MetaAddr> + '_ {
+    /// Return an iterator over peers that had recent attempts, successes, or failures,
+    /// in arbitrary order.
+    pub fn recently_used_peers(&'_ self) -> impl Iterator<Item = MetaAddr> + '_ {
         let _guard = self.span.enter();
 
-        self.by_addr
-            .values()
-            .filter(move |peer| self.maybe_connected_addr(&peer.addr))
+        self.peers_unordered()
+            .filter(move |peer| self.recently_used_addr(peer.addr))
             .cloned()
     }
 
-    /// Return an iterator over peers we've seen recently, in arbitrary order.
-    pub fn recently_live_peers(&'_ self) -> impl Iterator<Item = MetaAddr> + '_ {
+    /// Return an iterator over candidate peers, in arbitrary order.
+    ///
+    /// Candidate peers have not had recent attempts, successes, or failures.
+    fn candidate_peers(&'_ self) -> impl Iterator<Item = MetaAddr> + '_ {
         let _guard = self.span.enter();
 
-        self.by_addr
-            .values()
-            .filter(move |peer| self.recently_live_addr(&peer.addr))
+        // Skip recently used peers (including live peers)
+        self.peers_unordered()
+            .filter(move |peer| !self.recently_used_addr(peer.addr))
             .cloned()
     }
 
-    /// Returns an iterator that drains entries from the address book.
+    /// Return the next peer that is due for a connection attempt.
+    pub fn next_candidate_peer(&self) -> Option<MetaAddr> {
+        let _guard = self.span.enter();
+
+        self.candidate_peers().min()
+    }
```

Comment:
Security: Limit reconnection rate to each individual peer address #1848

Part 2: Ignore recently used peers when choosing candidate peers